app/src/main/java/org/thoughtcrime/securesms/video/videoconverter/VideoTrackConverter.java

```diff
@@ -0,0 +1,512 @@
+        } catch (Exception e) {
+            Log.e(TAG, "error while releasing mInputSurface", e);
+            if (exception == null) {
+                exception = e;
+            }
+        }
+        try {
+            if (mVideoEncoder != null) {
+                mVideoEncoder.stop();
+                mVideoEncoder.release();
+            }
+        } catch (Exception e) {
+            Log.e(TAG, "error while releasing mVideoEncoder", e);
+            if (exception == null) {
+                exception = e;
+            }
+        }
+        if (exception != null) {
+            throw exception;
+        }
+    }
+
+    String dumpState() {
+        return String.format(Locale.US,
+                "V{"
+                        + "extracted:%d(done:%b) "
+                        + "decoded:%d(done:%b) "
+                        + "encoded:%d(done:%b) "
+                        + "muxing:%b(track:%d)} ",
+                mVideoExtractedFrameCount, mVideoExtractorDone,
+                mVideoDecodedFrameCount, mVideoDecoderDone,
+                mVideoEncodedFrameCount, mVideoEncoderDone,
+                mMuxer != null, mOutputVideoTrack);
+    }
+
+    void verifyEndState() {
+        Preconditions.checkState("encoded (" + mVideoEncodedFrameCount + ") and decoded (" + mVideoDecodedFrameCount + ") video frame counts should match", mVideoDecodedFrameCount == mVideoEncodedFrameCount);
+        Preconditions.checkState("decoded frame count should be less than extracted frame count", mVideoDecodedFrameCount <= mVideoExtractedFrameCount);
+    }
+
+    private static String createFragmentShader(
+            final int srcWidth,
+            final int srcHeight,
+            final int dstWidth,
+            final int dstHeight) {
+        final float kernelSizeX = (float) srcWidth / (float) dstWidth;
+        final float kernelSizeY = (float) srcHeight / (float) dstHeight;
+        Log.i(TAG, "kernel " + kernelSizeX + "x" + kernelSizeY);
+        final String shader;
+        if (kernelSizeX <= 2 && kernelSizeY <= 2) {
+            shader =
+                    "#extension GL_OES_EGL_image_external : require\n" +
+                            "precision mediump float;\n" +      // highp here doesn't seem to matter
+                            "varying vec2 vTextureCoord;\n" +
+                            "uniform samplerExternalOES sTexture;\n" +
+                            "void main() {\n" +
+                            "    gl_FragColor = texture2D(sTexture, vTextureCoord);\n" +
+                            "}\n";
+        } else {
+            final int kernelRadiusX = (int) Math.ceil(kernelSizeX - .1f) / 2;
+            final int kernelRadiusY = (int) Math.ceil(kernelSizeY - .1f) / 2;
+            final float stepX = kernelSizeX / (1 + 2 * kernelRadiusX) * (1f / srcWidth);
+            final float stepY = kernelSizeY / (1 + 2 * kernelRadiusY) * (1f / srcHeight);
+            final float sum = (1 + 2 * kernelRadiusX) * (1 + 2 * kernelRadiusY);

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).